### PR TITLE
feat(availability): saves in 30 min chunks

### PR DESCRIPTION
**Pull request by esparr**
### **1. Targeted Issue** ###

Issue #121  https://trello.com/c/Jj3sK3SO

### **2. Overview of Solution** ###

Availability v2 endpoint takes in a date range and bulk creates availabilities in 30 minute chunks. The Availability model also now includes a status field with the choices: Open, Requested, Confirmed. 

- Front end will need to pass intervals of :00 and :30. 
- If there is time that is not in 30 min intervals, the endpoint will truncate the range and return pks that are in 30 mins intervals from the start time. 
Screen Shot 2023-10-24 at 12 57 24 PM

### **3. Tools Used** ###

- We are implementing a new versioning strategy recommended by Django Rest Framework using headers to pass along versioning: https://www.django-rest-framework.org/api-guide/versioning/#acceptheaderversioning

### **4. Testing Strategy** ###

I added coverage for all availability endpoints. Testing with new sessions will have to happen in the future. 

### **5. Future Implications** ###

This endpoint will create a lot of other tasks that will need to be implemented. 
- We will need to create a V2 for sessions. It will need to take in 1 to 2 availability objects, create the session and then mark availabilities requested or confirmed.
- We are implementing a new versioning strategy recommended by Django Rest Framework using headers to pass along versioning: https://www.django-rest-framework.org/api-guide/versioning/#acceptheaderversioning
- Times will need to be sent in 30 min rounded times (ex: 3pm, 3:30pm). Validation will need to happen on the front end first. 
- ***We need to figure how how we want to limit the total time range input at a time.***

### **6. Screenshots** ###

All Tests Passing
Screen Shot 2023-10-18 at 1 25 49 PM

Coverage Report for Availabilities
Screen Shot 2023-10-18 at 1 26 17 PM

Headers Set up
Screen Shot 2023-10-21 at 8 00 50 PM

Get Request and Response
Screen Shot 2023-10-21 at 8 02 08 PM

Post Request and Response
Screen Shot 2023-10-21 at 8 03 13 PM

### **7. Code Reviewers** ###

@jimbandrews @GitLukeW @nzeager 

---

Before you go...

- [x]  I have linked the relevant issue
- [x]  I have provided an overview of my solution
- [x]  I have shared the tools I've used
- [x]  I have outlined my testing strategy
- [x]  I have considered future implications of my changes
- [x]  I have included relevant screenshots
- [x]  I have kept my team in the loop

---

### CODE REVIEW

File: team_production_system/helpers.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    """
+    return start_time > timezone.now()
+
+
+# Check if the end_time is after the start_time
+def is_valid_end_time(start_time, end_time):
+    """
+    Check that the end_time is after the start_time.
+    """
+    if end_time > start_time:
```
Comment (jimbandrews) [inline team_production_system/helpers.py:59]: This should just return the inequality. The if-else is superfluous.

File: team_production_system/models/availability.py [hunk 2]
```diff
@@ -5,16 +5,28 @@
...
+    created_at = models.DateTimeField(auto_now_add=True)
+    modified_at = models.DateTimeField(auto_now=True)
 
     class Meta:
         constraints = [
             UniqueConstraint(
                 fields=['mentor', 'start_time'],
-                name='availability_constraint')
+                name='availability_constraint',
+                violation_error_message='Availability already exists.')
```
Comment (jimbandrews) [inline team_production_system/models/availability.py:29]: Nit-pick: Closing parentheses should get it's own line (we could probably configure the linter to do this for us)

File: team_production_system/models/availability.py [hunk 3]
```diff
@@ -5,16 +5,28 @@
...
 class Availability(models.Model):
+
+    STATUS_CHOICES = [
+        ('Open', 'Open'),
+        ('Requested', 'Requested'),
+        ('Confirmed', 'Confirmed'),
+    ]
+
     mentor = models.ForeignKey(
         Mentor, on_delete=models.CASCADE, related_name='mentor_availability')
```
Comment (jimbandrews) [inline team_production_system/models/availability.py:16]: Nit-pick: Closing parentheses should get it's own line

File: team_production_system/models/availability.py [hunk 4]
```diff
@@ -5,16 +5,28 @@
...
     start_time = models.DateTimeField()
     end_time = models.DateTimeField()
+    status = models.CharField(
+        max_length=10, choices=STATUS_CHOICES, default='Open')
```
Comment (jimbandrews) [inline team_production_system/models/availability.py:20]: Nit-pick: Each argument should get it's own line, as well as the closing parenthesis

File: team_production_system/serializers/availability.py [hunk 5]
```diff
@@ -79,4 +85,32 @@ def validate_end_time(self, value):
...
 
-    # TODO: Add validation for start times
+
+# V2 API #
+# The mentor availability serializer
+class AvailabilitySerializerV2(serializers.ModelSerializer):
+
+    class Meta:
+        model = Availability
+        fields = ['pk', 'mentor', 'start_time', 'end_time', 'status']
```
Comment (jimbandrews) [inline team_production_system/serializers/availability.py:95]: Nit-pick: each field should probably get it's own line to increase readability
```suggestion
        fields = [
            'pk',
            'mentor',
            'start_time',
            'end_time',
            'status'
        ]
```

File: team_production_system/serializers/availability.py [hunk 6]
```diff
@@ -79,4 +85,32 @@ def validate_end_time(self, value):
...
+        read_only_fields = ('pk',)
```
Comment (jimbandrews) [inline team_production_system/serializers/availability.py:96]: Does the comma need to be there?

File: team_production_system/serializers/availability.py [hunk 7]
```diff
@@ -79,4 +85,32 @@ def validate_end_time(self, value):
...
+
+    def validate(self, data):
```
Comment (jimbandrews) [inline team_production_system/serializers/availability.py:98]: How is this being called? I'm not calling for a fix, this is just a gap in my Django knowledge.

File: team_production_system/views/availability_list_create.py [hunk 8]
```diff
@@ -11,10 +16,57 @@ class AvailabilityListCreateView(generics.ListCreateAPIView):
...
+            serializer = self.get_serializer(data=availabilities, many=True)
+            serializer.is_valid(raise_exception=True)
+            self.perform_create(serializer)
+
+            headers = self.get_success_headers(serializer.data)
+            return Response(
+                serializer.data,
+                status=status.HTTP_201_CREATED,
+                headers=headers)
+        else:
```
Comment (jimbandrews) [inline team_production_system/views/availability_list_create.py:58]: Do you need the else line if you already have a return in the if statement?

**Review by jimbandrews — CHANGES_REQUESTED** [review]
Looks good! It seems like the versioning was pretty easy to implement. Most of my comments are just formatting nitpicks. I should probably look into how to add appropriate rules for the linter to pick them up.

File: team_production_system/models/availability.py (hunk 2 above)
Comment (esparr) [reply team_production_system/models/availability.py:29]: That is a good plan. I am down to use that formatting.

File: team_production_system/serializers/availability.py (hunk 7 above)
Comment (esparr) [reply team_production_system/serializers/availability.py:98]: Good question! DRF has a validate built-in method that picks this up. I am overriding part of it here.

**Review by GitLukeW — APPROVED** [review]
Thank you so much! Can't wait to see it in action!

**Pull request merged**